primefaces/src/main/java/org/primefaces/util/FileUploadUtils.java

```diff
@@ -224,111 +222,105 @@ protected static String convertJavaScriptRegex(String jsRegex) {
         }
         char endChar = jsRegex.charAt(end);
-        if (endChar != '/' && endChar == 'i' || endChar == 'g') {
+        if (endChar == 'i' || endChar == 'g') {
             end = end - 1;
         }
         return LangUtils.substring(jsRegex, start, end);
     }
 
-    private static boolean isValidFileContent(PrimeApplicationContext context, FileUpload fileUpload, String fileName, InputStream stream) throws IOException {
+    private static boolean isValidFileContent(PrimeApplicationContext primeAppContext, FileUpload fileUpload, String fileName, InputStream stream)
+            throws IOException {
         if (!fileUpload.isValidateContentType()) {
             if (LOGGER.isLoggable(Level.FINE)) {
                 LOGGER.fine("Content type checking is disabled");
             }
             return true;
         }
         if (LangUtils.isBlank(fileUpload.getAccept())) {
-            //Short circuit
             return true;
         }
 
-        String tempFilePrefix = UUID.randomUUID().toString();
-        Path tempFile = Files.createTempFile(tempFilePrefix, "");
-
-        try {
-            try (InputStream in = new PushbackInputStream(new BufferedInputStream(stream))) {
-                try (OutputStream out = new FileOutputStream(tempFile.toFile())) {
-                    IOUtils.copyLarge(in, out);
-                }
-            }
+        String prefix = FilenameUtils.removeExtension(fileName);
+        Path tempFile = Files.createTempFile(prefix, Constants.EMPTY_STRING);
+        String contentType = null;
 
+        try (InputStream in = new PushbackInputStream(new BufferedInputStream(stream));
+             OutputStream out = Files.newOutputStream(tempFile)) {
+            IOUtils.copyLarge(in, out);
 
-            String contentType = context.getFileTypeDetector().probeContentType(tempFile);
+            contentType = primeAppContext.getFileTypeDetector().probeContentType(tempFile);
 
             // We want to force the FileTypeDetector impl to try to detect the file type by its content,
             // this is why we intentionally hide the original file name/extension at first because both
             // tika and mime-types go the easy way of looking up the file extension if possible.
             // If this first attempt failed we try again, but now while preserving the original file name/extension
             // to e.g. make the JDK default FileTypeDetector work
             if (contentType == null) {
-                String fileExtension = FilenameUtils.getExtension(fileName);
-
-                if (!fileExtension.isEmpty()) {
-                    String newFileName = tempFile.getFileName().toString() + "." + fileExtension;
-                    tempFile = Files.move(tempFile, tempFile.resolveSibling(newFileName));
-
-                    contentType = context.getFileTypeDetector().probeContentType(tempFile);
-                }
+                tempFile = Files.move(tempFile, tempFile.resolveSibling(fileName));
```

Comment:
I think this could cause a collision if two files with the same name are uploaded simultaniously:
1. `test.jpg` is uploaded
2. It gets saved in a temporary file with a randomized name, e.g. `/tmp/test3887755495119540932` (fine)
3. Then it's moved according to its original filename: `/tmp/test.jpg` (problematic, could exist already because of another request, some other program writing temporary files, etc.)

Really what i think should happen is that the tempFile name is taken as is (we can expect it to be unique because of the random name) and the file extension gets appended